include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -89,10 +89,16 @@ class ApplyWithInitialGuess {
      * @param x  the input vectors
      * @param guess  the input guess
      */
-    template <typename ValueType>
-    static void prepare_initial_guess(const matrix::Dense<ValueType>* b,
-                                      matrix::Dense<ValueType>* x,
-                                      initial_guess_mode guess);
+    template <typename VectorType>
+    static void prepare_initial_guess(const VectorType* b, VectorType* x,
+                                      initial_guess_mode guess)
+    {
+        if (guess == initial_guess_mode::zero) {
+            x->fill(zero<typename VectorType::value_type>());
+        } else if (guess == initial_guess_mode::rhs) {
+            x->copy_from(b);
+        }
+    }
```

Comment:
The interface should only provide publically available operations. This looks more like a free function, which could probably move to a private header to reduce the amount of functions we freeze into the interface.

Reply:
your are right